lib/rpmfi.c

```diff
@@ -1508,53 +1512,59 @@ static void rpmfilesBuildNLink(rpmfiles fi, Header h)
     return;
 }
 
-/* Convert a tag of hex strings to binary presentation */
-/* If lengths is non-NULL, assume variable length strings */
-static uint8_t *hex2bin(Header h, rpmTagVal tag, rpm_count_t num, size_t len,
-			int **lengths, int *maxlen)
+/*
+ * Convert a tag of variable len hex strings to binary presentation,
+ * accessed via offsets to a contiguous binary blob. Empty values
+ * are represented by identical consequtive offsets. The offsets array
+ * always has one extra element to allow calculating the size of the
+ * last element.
+ */
+static uint8_t *hex2binv(Header h, rpmTagVal tag, rpm_count_t num,
+			uint32_t **offsetp)
 {
     struct rpmtd_s td;
     uint8_t *bin = NULL;
+    uint32_t *offs = NULL;
 
     if (headerGet(h, tag, &td, HEADERGET_MINMEM) && rpmtdCount(&td) == num) {
 	const char *s;
-	int maxl = 0;
-	int *lens = NULL;
-
-	/* Figure string sizes + max length for allocation purposes */
-	if (lengths) {
-	    int i = 0;
-	    lens = xmalloc(num * sizeof(*lens));
-
-	    while ((s = rpmtdNextString(&td))) {
-		lens[i] = strlen(s) / 2;
-		if (lens[i] > maxl)
-		    maxl = lens[i];
-		i++;
-	    }
-
-	    *lengths = lens;
-	    *maxlen = maxl;
+	int i = 0;
+	uint8_t *t = bin = xmalloc(((rpmtdSize(&td) / 2) + 1));
+	offs = xmalloc((num + 1) * sizeof(*offs));
 
-	    /* Reinitialize iterator for next round */
-	    rpmtdInit(&td);
-	} else {
-	    maxl = len;
+	while ((s = rpmtdNextString(&td))) {
+	    uint32_t len = strlen(s) / 2;
+	    offs[i] = t - bin;
+	    for (int j = 0; j < len; j++, t++, s += 2)
+		*t = (rnibble(s[0]) << 4) | rnibble(s[1]);
+	    i++;
 	}
+	offs[i] = t - bin;
+	*offsetp = offs;
+
+	rpmtdFreeData(&td);
+    }
+    return bin;
+}
+
+/* Convert a tag of hex strings to binary presentation */
+static uint8_t *hex2bin(Header h, rpmTagVal tag, rpm_count_t num, size_t len)
+{
+    struct rpmtd_s td;
+    uint8_t *bin = NULL;
+
+    if (headerGet(h, tag, &td, HEADERGET_MINMEM) && rpmtdCount(&td) == num) {
+	uint8_t *t = bin = xmalloc(num * len);
+	const char *s;
 
-	uint8_t *t = bin = xmalloc(num * maxl);
-	int i = 0;
 	while ((s = rpmtdNextString(&td))) {
 	    if (*s == '\0') {
```

Comment:
```suggestion
	    if (strlen(s) != len * 2) {
```

This will ensure that any junk digests (that have incorrect lengths) are ignored (and then, hopefully, cause package install to fail because they do not verify).  `len` is trusted here because is the return value of `rpmDigestLength()`.